[SYCL] Support -ftarget-compile-fast in JIT mode

sarnex opened the pull request:
This option is currently only supported for AOT mode, and it's implemented in the driver by passing an
`-igc_opts` string to ocloc.

Support this in JIT to by passing the FE option to `clang-offload-wrapper` so it's stored in the binary and will get used at runtime.

At runtime, check with the backend to get the correct backend flag, and use that. If the device is not an Intel GPU, strip the FE flag.

sarnex commented:
@aelovikov-intel Could review my e2e test please (just the RUN commands), it's a little bit unusual

aelovikov-intel commented on sycl/test-e2e/KernelAndProgram/target_compile_fast.cpp:
```diff
@@ -0,0 +1,39 @@
+// RUN: env SYCL_PI_TRACE=-1 %{run} %t_with.out 2>&1 | FileCheck %if !gpu || ext_oneapi_hip || ext_oneapi_cuda %{ --implicit-check-not=-igc_opts %} %else %{ --check-prefix=CHECK-INTEL-GPU-WITH %} %s
+// RUN: env SYCL_PI_TRACE=-1 %{run} %t_without.out 2>&1 | FileCheck --implicit-check-not=-igc_opts %s
+
+// CHECK-INTEL-GPU-WITH: ---> piProgramBuild(
+// CHECK-INTEL-GPU-WITH: -igc_opts 'PartitionUnit=1,SubroutineThreshold=50000'
```
I don't think anything prevents this from running on cpu/accelerator/CUDA/HIP at the moment.

mdtoguchi reviewed (approved):
Looks OK for Driver

sarnex commented:
Ping @intel/llvm-reviewers-runtime for this one, thanks!

sarnex commented:
@steffenlarsen @dm-vodopyanov Any interest in reviewing? Thanks

steffenlarsen commented:
> I know we are moving to the unified runtime, so please let me know if I need to make any changes to support that.

Tag @smaslov-intel

smaslov-intel commented:
> > I know we are moving to the unified runtime, so please let me know if I need to make any changes to support that.
> 
> Tag @smaslov-intel

Why is this not using `piPluginGetBackendOption` so that each platform returns the options supported by underlying backend? See https://github.com/intel/llvm/blob/c6500e41fdc02545ae1867e9c3a868734ecc62c2/sycl/include/sycl/detail/pi.h#L2085

sarnex commented:
> Why is this not using piPluginGetBackendOption so that each platform returns the options supported by underlying backend? 

Because I didn't know it existed :) let me add this, thx

sarnex commented:
@smaslov-intel addressed your feedback in latest commit hopefully

@mdtoguchi i changed the implementation based on feedback, will need re-review, thanks

sarnex commented:
Failures not related.

sarnex commented:
@againull hey, do you mind reviewing this while sergey is OOO? thanks!

againull commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -457,17 +457,33 @@ static void appendCompileOptionsFromImage(std::string &CompileOpts,
+    if (auto Pos = CompileOpts.find(TargetCompileFast);
+        Pos != std::string::npos) {
+      const char *BackendOption = nullptr;
+      PlatformImpl->getBackendOption(TargetCompileFast, &BackendOption);
+      auto OptLen = strlen(TargetCompileFast);
+      if (IsIntelGPU && BackendOption && BackendOption[0] != '\0')
+        CompileOpts.replace(Pos, OptLen, BackendOption);
+      else
+        CompileOpts.erase(Pos, OptLen);
```
It looks like your intention was to ignore this option if backend is not an Intel GPU.
But when you call getBackendOption for CUDA/HIP plugins it will result in exception because those plugins don't know about this new option.

againull reviewed (commented):
It will be nice to document this experimental option  in https://github.com/intel/llvm/blob/sycl/sycl/doc/UsersManual.md

sarnex commented:
@intel/dpcpp-doc-reviewers ping on this one? Just documented a new experimental option that we are supporting in JIT mode. The option was already added for AOT mode, we just didn't document it.

bader commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -991,10 +990,13 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
     if (Args.hasArg(options::OPT_ftarget_compile_fast)) {
       BeArgs.push_back("-igc_opts 'PartitionUnit=1,SubroutineThreshold=50000'");
     }
+  } else if (!Triple.isNVPTX() && !Triple.isAMDGCN()) {
```
The documentation says "Currently only supported on Intel GPUs.", but I read this as "Currently everything except NVPTX and AMDGDN targets are supported".
Is it possible to replace this check with Intel GPU (i.e. SPIR-V) target?

bader replied:
@sarnex, how does adding `Triple.getSubArch() == llvm::Triple::NoSubArch` resolve the comment above?

bader replied:
@sarnex, ping.

bader replied:
> @bader this is JIT, so we don't know what the hardware is in the driver, we can't check for intel gpu jit here. the NVPTX and AMDGCN checks are to prevent us to go in this if for AOT with those targets. intel gpu aot (gen) is checked above, and the other jit spir subtargets are fpga and cpu, so checking nosubarch makes sure we don't go in for those two cases. did i misunderstand your comment?

I think we better check for one supported target. Ruling out unsupported targets doesn't scale.

sarnex replied:
good point, thanks, will do

sarnex commented:
Sorry for force-push, there was some CI problem

sarnex commented:
@bader Can you re-review, thanks!

sarnex commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -991,10 +990,14 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
     if (Args.hasArg(options::OPT_ftarget_compile_fast)) {
       BeArgs.push_back("-igc_opts 'PartitionUnit=1,SubroutineThreshold=50000'");
     }
+  } else if (Triple.getSubArch() == llvm::Triple::NoSubArch &&
+             !Triple.isNVPTX() && !Triple.isAMDGCN()) {
```
@mdtoguchi will the above suggestion make sure we don't go into the branch for NVPTX/AMDGCN? Thx

mdtoguchi replied:
@sarnex, yes.  `nvptx` and `amdgcn` do not set `spir`

sarnex replied:
thanks mike!

bader commented:
@sarnex, please, resolve merge conflicts.

sarnex commented:
@bader all feedback should be addressed now, sorry for the noise with the pending comments

sarnex commented:
@intel/llvm-gatekeepers This one is ready for merge, thanks!
